state/cloudservice.go

```diff
@@ -21,60 +21,78 @@ type CloudService interface {
 	Addresses() []network.Address
 }
 
-// cloudService is an implementation of CloudService.
-type cloudService struct {
+// CloudService is an implementation of CloudService.
+type CloudService struct {
+	st  *State
 	doc cloudServiceDoc
 }
 
 type cloudServiceDoc struct {
-	// Id holds cloud service document key.
-	// It is the global key of the application represented
-	// by this service.
-	Id string `bson:"_id"`
+	// DocID holds cloud service document key.
+	DocID string `bson:"_id"`
 
 	ProviderId string    `bson:"provider-id"`
 	Addresses  []address `bson:"addresses"`
 }
 
+func newCloudService(st *State, doc *cloudServiceDoc) *CloudService {
+	svc := &CloudService{
+		st:  st,
+		doc: *doc,
+	}
+	return svc
+}
+
 // Id implements CloudService.
-func (c *cloudService) Id() string {
-	return c.doc.Id
+func (c *CloudService) Id() string {
+	return c.doc.DocID
 }
 
 // ProviderId implements CloudService.
-func (c *cloudService) ProviderId() string {
+func (c *CloudService) ProviderId() string {
 	return c.doc.ProviderId
 }
 
-// Address implements CloudService.
-func (c *cloudService) Addresses() []network.Address {
+// Addresses implements CloudService.
+func (c *CloudService) Addresses() []network.Address {
 	return networkAddresses(c.doc.Addresses)
 }
 
-func (a *Application) cloudService() (*cloudServiceDoc, error) {
-	coll, closer := a.st.db().GetCollection(cloudServicesC)
+func (c *CloudService) cloudService() (cloudServiceDoc, error) {
```

Comment:
can we continue to return a pointer *cloudServiceDoc

Reply:
changed